tcmalloc/internal/percpu_rseq_aarch64.S

```diff
@@ -0,0 +1,481 @@
+  .quad .L##label##_commit - .L##label##_start;                   \
+  .quad label##_trampoline;                                       \
+  PINSECTION(.L##label##array);                                   \
+  .popsection;                                                    \
+  .pushsection __rseq_cs_ptr_array, "aw";                         \
+  .L##label##array:                                               \
+  .quad __rseq_cs_##label;                                        \
+  .popsection;                                                    \
+  .pushsection rseq_trampoline, "ax";                             \
+  SIGN_ABORT();                                                   \
+  .globl label##_trampoline;                                      \
+  .type  label##_trampoline, @function;                           \
+label##_trampoline:                                               \
+  .cfi_startproc;                                                 \
+  b .L##label##_abort;                                            \
+  .cfi_endproc;                                                   \
+  .size label##_trampoline, . - label##_trampoline;               \
+  .popsection;
+
+// This is part of the upstream rseq ABI.  The 4 bytes prior to the abort IP
+// must match TCMALLOC_PERCPU_RSEQ_SIGNATURE (as configured by our rseq
+// syscall's signature parameter).  This signature is used to annotate valid
+// abort IPs (since rseq_cs could live in a user-writable segment).
+// We use .inst here instead of a data directive so it works for both small and
+// big endian.
+#define SIGN_ABORT()           \
+  .inst TCMALLOC_PERCPU_RSEQ_SIGNATURE
+
+/*
+ * Provide a directive to specify the size of symbol "label", relative to the
+ * current location and its start.
+ */
+#define ENCODE_SIZE(label) .size label, . - label
+/* We are assuming small memory model.  */
+#if __clang_major__ >= 9 && !defined(__AARCH64_CMODEL_SMALL__)
+#error "Memory model not supported!"
+#endif
+
+/* FETCH_CPU assumes &__rseq_abi is in x5.  */
+#define FETCH_CPU(dest) \
+  ldr dest, [x5, #4] /* cpuid is 32-bits */
+
+/* With PIE  have initial-exec TLS, even in the presence of position
+   independent code. */
+#if !defined(__PIC__) || defined(__PIE__)
+
+#define START_RSEQ(src)                         \
+  .L##src##_abort:                              \
+  mrs     x5, tpidr_el0;                        \
+  adrp    x6, :gottprel:__rseq_abi;             \
+  ldr     x6, [x6,:gottprel_lo12:__rseq_abi];   \
+  add     x5, x5, x6;                           \
+  adrp    x6, __rseq_cs_##src;                  \
+  add     x6, x6, :lo12:__rseq_cs_##src;        \
+  str     x6, [x5, #8];                         \
+  .L##src##_start:
+
+#else  /* !defined(__PIC__) || defined(__PIE__) */
+
+/*
+ * In the case where we can't guarantee we have initial-exec TLS we use
+ * the General Dynamic TLS model to get the address of __rseq_abi.
```

Comment:
The comment about "General Dynamic" is not correct. On AArch64 TLS descriptor is used, which is different from traditional General Dynamic/Local Dynamic TLS models.

Reply:
Thanks for raising this. I believe I crossed my terminology there, the TLS 'dialect' I use here is indeed TLS descriptor and not traditional TLS.

Would you be ok with:
"In the case where we can't guarantee we have initial-exec TLS we obtain __rseq_abi's TP offset using a TLS descriptor sequence, which we then add to the TP to get __rseq_abi's address.
The call to the TLS descriptor can be optimized away by the linker (keep the rest the same...)"

Reply:
LGTM. Thanks!